lib/internal/fs/streams.js

```diff
@@ -386,22 +386,71 @@ WriteStream.prototype.open = openWriteFs;
 
 WriteStream.prototype._construct = _construct;
 
+function writeAll(data, size, pos, cb, retries = 0) {
+  this[kFs].write(this.fd, data, 0, size, pos, (er, bytesWritten, buffer) => {
+    if (er?.code === 'EAGAIN') {
+      er = null;
+      bytesWritten = 0;
+    }
+
+    if (this.destroyed || er) {
+      return cb(er);
+    }
+
+    this.bytesWritten += bytesWritten;
+
+    retries = bytesWritten ? 0 : retries + 1;
+    size -= bytesWritten;
+    pos += bytesWritten;
+
+    if (retries > 5) {
+      cb(new Error('writev failed'));
+    } else if (size) {
+      writeAll(buffer.slice(bytesWritten), size, pos, cb, retries);
+    } else {
+      cb();
+    }
+  });
+}
+
+function writevAll(chunks, size, pos, cb, retries = 0) {
+  this[kFs].writev(this.fd, chunks, this.pos, (er, bytesWritten, buffers) => {
+    if (er?.code === 'EAGAIN') {
+      er = null;
+      bytesWritten = 0;
+    }
+
+    if (this.destroyed || er) {
+      return cb(er);
+    }
+
+    this.bytesWritten += bytesWritten;
+
+    retries = bytesWritten ? 0 : retries + 1;
+    size -= bytesWritten;
+    pos += bytesWritten;
+
+    if (retries > 5) {
+      cb(new Error('writev failed'));
+    } else if (size) {
+      writevAll([Buffer.concat(buffers).slice(bytesWritten)], size, pos, cb, retries);
```

Comment:
If we're going to be concatenating, we may as well just do it once and call `writeAll()` with it instead. Otherwise like @mcollina said, we should handle this more efficiently.

Ideally it would be great if we didn't have to even keep recreating arrays on retry. That might be a nice addition to the `fs.writev()` API to make it more like `fs.write()` with its `offset` parameter, so we could just pass a starting index or something.

Reply:
This is a cold path though? It should almost never happen. Is it worth optimizing for?

Reply:
> It should almost never happen

It think it would be useful to get an idea of how often it happens (I personally have no idea). How cold is it?

Reply:
Well, if it does happen often then we have a lot of seriously broken software out there... I guess we would have quite a bit of reports....

Reply:
or people have learned to handle it on their own?

Reply:
> Well, if it does happen often then we have a lot of seriously broken software out there... I guess we would have quite a bit of reports....

makes sense, yeah sounds like a cold path to me.

Reply:
Concatenating like this will fail if the buffers collectively exceed kMaxLength, though. Not just a perf issue.